src/inmanta/data/dataview.py

```diff
@@ -0,0 +1,1163 @@
+    DesiredStateVersionOrder,
+    FactOrder,
+    InvalidQueryParameter,
+    InvalidSort,
+    Notification,
+    NotificationOrder,
+    PagingOrder,
+    Parameter,
+    ParameterOrder,
+    QueryFilter,
+    ResourceAction,
+    ResourceHistoryOrder,
+    ResourceLogOrder,
+    ResourceOrder,
+    SimpleQueryBuilder,
+    VersionedResourceOrder,
+    model,
+)
+from inmanta.data.model import (
+    BaseModel,
+    CompileReport,
+    DesiredStateLabel,
+    DesiredStateVersion,
+    Fact,
+    LatestReleasedResource,
+    PagingBoundaries,
+    ResourceHistory,
+    ResourceIdStr,
+    ResourceLog,
+    ResourceVersionIdStr,
+)
+from inmanta.protocol.exceptions import BadRequest
+from inmanta.protocol.return_value_meta import ReturnValueWithMeta
+from inmanta.resources import Id
+from inmanta.server.validate_filter import (
+    BooleanEqualityFilter,
+    BooleanIsNotNullFilter,
+    CombinedContainsFilterResourceState,
+    ContainsFilter,
+    ContainsFilterResourceAction,
+    ContainsPartialFilter,
+    DateRangeFilter,
+    Filter,
+    FilterValidator,
+    IntRangeFilter,
+    InvalidFilter,
+    LogLevelFilter,
+)
+from inmanta.types import SimpleTypes
+from inmanta.util import datetime_utc_isoformat
+
+T_ORDER = TypeVar("T_ORDER", bound=DatabaseOrderV2)
+T_DTO = TypeVar("T_DTO", bound=BaseModel)
+
+
+class RequestedPagingBoundaries:
+    """Represents the lower and upper bounds that the user requested for the paging boundaries"""
+
+    def __init__(
+        self,
+        start: Optional[PRIMITIVE_SQL_TYPES],
+        end: Optional[PRIMITIVE_SQL_TYPES],
+        first_id: Optional[PRIMITIVE_SQL_TYPES],
+        last_id: Optional[PRIMITIVE_SQL_TYPES],
+        id_nullable: bool,
+    ) -> None:
+        self.start = start
+        self.end = end
+        self.first_id = first_id
+        self.last_id = last_id
+        self.id_nullable = id_nullable
+
+    def validate(
+        self,
+    ) -> None:
+        start = self.start
+        end = self.end
+        first_id = self.first_id
+        last_id = self.last_id
+        if (start is not None) and (end is not None):
+            raise InvalidQueryParameter(
+                f"Only one of start and end parameters is allowed at the same time. Received start: {start}, end: {end}"
+            )
+        if (first_id is not None) and (last_id is not None):
+            raise InvalidQueryParameter(
+                f"Only one of first_id and last_id parameters is allowed at the same time. "
+                f"Received first_id: {first_id}, last_id: {last_id}"
+            )
+
+        if self.has_start() and self.has_end():
+            raise InvalidQueryParameter(
+                f"Start and end parameters can not be set at the same time: "
+                f"Received first_id: {first_id}, last_id: {last_id}, start: {start}, end: {end}"
+            )
+
+        if self.id_nullable:
+            return
+
+        if (first_id and not start) or (first_id and end):
```

Comment:
```suggestion
        if (first_id and not start):
```
I think this second condition was already covered by `if self.has_start() and self.has_end()`.